clients/client-py/test/test_upload_download.py

```diff
@@ -0,0 +1,209 @@
+            },
+        }
+
+    def finishUpload(self, name, payload):
+        assert payload["uploadId"] == self.lastUploadId
+        assert payload["projectId"] == self.lastProjectId
+
+        return {}
+
+
+def test_simple_download_fails():
+    "When a simple download's GET fails with a 400, an exception is raised and no retries occur"
+    getcount = 0
+
+    class Server(httptest.Handler):
+        def do_GET(self):
+            nonlocal getcount
+            getcount += 1
+            self.send_response(400)
+            self.end_headers()
+            self.wfile.write(b'uhoh')
+
+    with httptest.Server(Server) as ts:
+        objectService = FakeObject(ts)
+        with pytest.raises(requests.RequestException):
+            download.downloadToBuf(
+                name="some/object",
+                objectService=objectService)
+        assert getcount == 1
+
+
+def test_simple_download_fails_retried():
+    "When a simple download's GET fails with a 500, an exception is raised after five retries"
+    attempts = 0
+
+    class Server(httptest.Handler):
+        def do_GET(self):
+            nonlocal attempts
+            attempts += 1
+            self.send_response(500)
+            self.end_headers()
+            self.wfile.write(b'uhoh')
+
+    with httptest.Server(Server) as ts:
+        objectService = FakeObject(ts)
+        with pytest.raises(requests.RequestException):
+            download.downloadToBuf(
+                name="some/object",
+                objectService=objectService)
+
+    assert attempts == 6  # one try plus five retries
+
+
+def test_simple_download_fails_retried_succeeds(randbytes):
+    "When a simple download's GET fails with a 500, it is retried successfully"
+    attempts = 0
+    data = randbytes(1024)
+
+    class Server(httptest.Handler):
+        def do_GET(self):
+            nonlocal attempts
+            attempts += 1
+            if attempts > 2:
+                self.send_response(200)
+                self.send_header('content-type', 'text/plain')
+                self.send_header('content-length', str(len(data)))
+                self.end_headers()
+                self.wfile.write(data)
+            else:
+                self.send_response(500)
+                self.end_headers()
+                self.wfile.write(b'uhoh')
+
+    with httptest.Server(Server) as ts:
+        objectService = FakeObject(ts)
+        buf, content_type = download.downloadToBuf(
+            name="some/object",
+            objectService=objectService)
+
+    assert attempts == 3
+    assert buf == data
+    assert content_type == 'text/plain'
+
+
+def test_putUrl_upload_fails(randbytes):
+    "When a putUrl upload's PUT fails with a 400, an exception is raised"
+    data = randbytes(10240)  # >8k to avoid using dataInline
+
+    class Server(httptest.Handler):
+        def do_PUT(self):
+            self.send_response(400)
+            self.end_headers()
+            self.wfile.write(b'uhoh')
```

Comment:
Should this test method also have an ``attempts`` counter with an assertion? I think 400 is the case that fails without retries.

Reply:
Indeed, and it appears this is also buggy!

Reply:
I addressed this (it was the same bug as pete noted in `taskcluster/download.py`).